motionplan/cBiRRT.go

```diff
@@ -110,115 +105,167 @@ func NewCBiRRTMotionPlannerWithSeed(frame referenceframe.Frame, nCPU int, seed *
+		if time.Since(mp.start) > time.Duration(algOpts.Timeout)*time.Second {
+			solutionChan <- &rrtPlanReturn{err: errPlannerTimeout, rm: rm}
+			return
+		}
+
 		select {
 		case <-ctx.Done():
-			solutionChan <- &planReturn{err: ctx.Err()}
+			solutionChan <- &rrtPlanReturn{err: ctx.Err()}
 			return
 		default:
 		}
 
-		// attempt to extend map1 first
+		// attempt to extend maps 1 and 2 towards the target
 		nearest1 := nm.nearestNeighbor(nmContext, planOpts, target, map1)
-		map1reached := mp.constrainedExtend(ctx, algOpts, map1, nearest1, &basicNode{q: target})
-
-		// then attempt to extend map2 towards map 1
-		nearest2 := nm.nearestNeighbor(nmContext, planOpts, map1reached.Q(), map2)
-		map2reached := mp.constrainedExtend(ctx, algOpts, map2, nearest2, map1reached)
+		nearest2 := nm.nearestNeighbor(nmContext, planOpts, target, map2)
+		//nolint: gosec
+		rseed1 := rand.New(rand.NewSource(int64(mp.randseed.Int())))
+		//nolint: gosec
+		rseed2 := rand.New(rand.NewSource(int64(mp.randseed.Int())))
+
+		utils.PanicCapturingGo(func() {
+			mp.constrainedExtend(ctx, algOpts, rseed1, map1, nearest1, &basicNode{q: target}, m1chan)
+		})
+		utils.PanicCapturingGo(func() {
+			mp.constrainedExtend(ctx, algOpts, rseed2, map2, nearest2, &basicNode{q: target}, m2chan)
+		})
+		map1reached := <-m1chan
+		map2reached := <-m2chan
 
 		corners[map1reached] = true
 		corners[map2reached] = true
 
 		_, reachedDelta := planOpts.DistanceFunc(&ConstraintInput{StartInput: map1reached.Q(), EndInput: map2reached.Q()})
-		if reachedDelta < algOpts.JointSolveDist {
+
+		// Second iteration; extend maps 1 and 2 towards the halfway point between where they reached
+		if reachedDelta > algOpts.JointSolveDist {
+			target = referenceframe.InterpolateInputs(map1reached.Q(), map2reached.Q(), 0.5)
+
+			nearest1 := nm.nearestNeighbor(nmContext, planOpts, target, map1)
+			nearest2 := nm.nearestNeighbor(nmContext, planOpts, target, map2)
+
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, algOpts, rseed1, map1, nearest1, &basicNode{q: target}, m1chan)
+			})
+			utils.PanicCapturingGo(func() {
+				mp.constrainedExtend(ctx, algOpts, rseed2, map2, nearest2, &basicNode{q: target}, m2chan)
+			})
+			map1reached = <-m1chan
+			map2reached = <-m2chan
+			_, reachedDelta = planOpts.DistanceFunc(&ConstraintInput{StartInput: map1reached.Q(), EndInput: map2reached.Q()})
+		}
+
+		// Solved!
+		if reachedDelta <= algOpts.JointSolveDist {
 			cancel()
-			path := extractPath(seedMap, goalMap, &nodePair{map1reached, map2reached})
+			path := extractPath(rm.startMap, rm.goalMap, &nodePair{map1reached, map2reached})
 			if endpointPreview != nil {
 				endpointPreview <- path[len(path)-1]
 			}
 			finalSteps := mp.SmoothPath(ctx, algOpts, path, corners)
-			solutionChan <- &planReturn{steps: finalSteps}
+			solutionChan <- &rrtPlanReturn{steps: finalSteps}
 			return
 		}
 
 		// sample near map 1 and switch which map is which to keep adding to them even
```

Comment:
[nit] this isn't really true anymore, and we might not even need to swap the maps, but I suppose it doesn't hurt either